crates/containerd-shim-wasm/src/sandbox/shim/otel.rs

```diff
@@ -0,0 +1,164 @@
+//! OpenTelemetry Configuration Module
+//!
+//! This module provides a configuration structure and associated methods to initialize
+//! OpenTelemetry tracing with the OTLP exporter. The configuration can be set up via
+//! the `OtelConfig` struct and its builder pattern.
+//!
+//! # Usage
+//!
+//! ```rust
+//! use containerd-shim-wasm::sandbox::shim::otel::{OtelConfig, OTEL_EXPORTER_OTLP_ENDPOINT};
+//!
+//! fn main() -> anyhow::Result<()> {
+//!     let otel_endpoint = std::env::var(OTEL_EXPORTER_OTLP_ENDPOINT).expect("OTEL_EXPORTER_OTLP_ENDPOINT not set");
+//!     let otel_config = OtelConfig::builder()
+//!         .otel_endpoint(otel_endpoint)
+//!         .name("my-service".to_string())
+//!         .build()?;
+//!
+//!     let _guard = otel_config.init()?;
+//!
+//!     // Your application code here
+//!
+//!     Ok(())
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use opentelemetry::global::{self, set_text_map_propagator};
+use opentelemetry::trace::TraceError;
+use opentelemetry_otlp::{
+    SpanExporterBuilder, WithExportConfig, OTEL_EXPORTER_OTLP_PROTOCOL_DEFAULT,
+};
+pub use opentelemetry_otlp::{OTEL_EXPORTER_OTLP_ENDPOINT, OTEL_EXPORTER_OTLP_PROTOCOL};
+use opentelemetry_sdk::propagation::TraceContextPropagator;
+use opentelemetry_sdk::{runtime, trace as sdktrace};
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt as _;
+use tracing_subscriber::layer::SubscriberExt as _;
+use tracing_subscriber::{EnvFilter, Registry};
+
+const OTEL_EXPORTER_OTLP_PROTOCOL_HTTP_PROTOBUF: &str = "http/protobuf";
+const OTEL_EXPORTER_OTLP_PROTOCOL_GRPC: &str = "grpc";
+
+/// Configuration struct for OpenTelemetry setup.
+pub struct Config {
+    otel_endpoint: String,
+    otel_protocol: String,
+}
+
+/// Initializes a new OpenTelemetry tracer with the OTLP exporter.
+///
+/// Returns a `Result` containing the initialized tracer or a `TraceError` if initialization fails.
+///
+/// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options
+impl Config {
+    /// Creates a new builder for `OtelConfig`.
+    pub fn builder() -> ConfigBuilder {
+        ConfigBuilder::default()
+    }
+
+    /// Initializes the tracer, sets up the telemetry and subscriber layers, and sets the global subscriber.
+    pub fn init(&self) -> anyhow::Result<ShutdownGuard> {
+        let tracer = self.init_tracer()?;
+        let telemetry = tracing_opentelemetry::layer().with_tracer(tracer);
+        set_text_map_propagator(TraceContextPropagator::new());
+
+        let filter = EnvFilter::try_new("info,h2=off")?;
+
+        let subscriber = Registry::default().with(telemetry).with(filter);
+
+        tracing::subscriber::set_global_default(subscriber)?;
```

Comment:
Yeah that's a good point. Since both the shim and spin runtime are multi-threaded, I don't see a good solution here. If we defer setting the default to runtimes, then we will lose the traces for the shim functions. 

I am experimenting moving away from `tracing` library and use `opentelemetry-sdk` to instrument spans. In this way, we don't need to call `set_global_default` and can use named `tracer`.